File: algorithm/ultraxrcrypt.py
import random
import re
from math import sin

alphabet = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890"
start_marker = "{~xr~}+"
end_marker = "-{~xr~}"

def hash(key, alphabet=alphabet, length=64):
    key = str(key)
    result = ""
    keysum = sum([ord(c) for c in key])
    for i in range(length):
        keyitem = ord(key[i % len(key)])
        
        hasheditem = (keyitem*keyitem + sin(keyitem * i * i - 29) * 110 + 401) % 1000
        hasheditem += (keyitem*2 + sin(keyitem * i * 91 - keysum * 7) * 160 + 301) % 1000
        hasheditem += (keyitem*5 + sin(i * 13 * keysum * 3) * 102 + 201) % 777
        hasheditem += sin(keyitem * i + keysum * keyitem) * keysum + 101
        hasheditem += keyitem * i + 77 * (i % len(key))
        
        result += alphabet[int(hasheditem) % len(alphabet)]
    
    return result

def generate(length, limits):
    result = ""
    for i in range(length):
        result += chr(random.randint(limits[0], limits[1]))
    return result

def generate_noise(min_length=1, max_length=16, context=alphabet):
    if len(context) == 0:
        context = alphabet

    noise = ""
    length = random.randint(min_length, max_length)
    for i in range(length):
        noise += context[random.randint(0, len(context) - 1)]

    return start_marker + noise + end_marker

def add_noise(text, count, min_length=1, max_length=16, context=alphabet):
    selected = []
    textlen = len(text)
    for i in range(count):
        selected.append(random.randint(0, textlen - 1))
    selected.sort()
    result = ""
    length = len(text)
    for i in range(length):
        result += text[i]
        if i in selected:
            noise_to_add = generate_noise(min_length, max_length, context)
            result += noise_to_add
            length += len(noise_to_add)

    return result    

def noise_remover(text):
    return re.sub(r"\{~xr~\}\+.*?\-\{~xr~\}", "", text)

def uxrcrypt(text, key, padding=16):
    encrypted_result = ""

    text = add_noise(text, 5 + random.randint(1,11), 1, 16, alphabet)

    text = generate_noise(1, padding, alphabet) + text + generate_noise(1, padding, alphabet)

    keylen = len(key)
    textlen = len(text)
    keysum = sum([ord(c) for c in key])

    for i in range(textlen):
        keyitem = ord(key[i % keylen])
        encrypted_char = ord(text[i])
        
        encrypted_char += (keylen * keylen + keyitem * keyitem) % 111
        encrypted_char += (37 * keyitem) % 31
        encrypted_char += ((textlen * textlen + 7) * keylen) % 222
        encrypted_char += (keylen * textlen - 1) % 333
        encrypted_char += (keysum * textlen + keysum * keylen - keysum ^ keyitem) % 255
        encrypted_char -= ((2 * textlen)%8 + 1) % 33
        encrypted_char += ((int(i % keylen) * 3) % 37) % 555
        encrypted_char += ((keyitem * 11) % 5) * 1001 + 777
        
        try:
            encrypted_result += chr(int(encrypted_char))
        except:
            print("Couln't add character: " + str(encrypted_char))
            exit(1)

    return encrypted_result

def deuxrcrypt(text, key):
    decrypted_result = ""

    keylen = len(key)
    textlen = len(text)
    keysum = sum([ord(c) for c in key])

    for i in range(textlen):
        keyitem = ord(key[i % keylen])
        textitem = ord(text[i])
        decrypted_char = ord(text[i])
        
        decrypted_char -= ((keyitem * 11) % 5) * 1001 + 777
        decrypted_char -= ((int(i % keylen) * 3) % 37) % 555
        decrypted_char += ((2 * textlen)%8 + 1) % 33
        decrypted_char -= (keysum * textlen + keysum * keylen - keysum ^ keyitem) % 255
        decrypted_char -= ((keylen * textlen - 1) % 333)
        decrypted_char -= ((textlen * textlen + 7) * keylen) % 222
        decrypted_char -= (37 * keyitem) % 31
        decrypted_char -= (keylen * keylen + keyitem * keyitem) % 111
        
        try:
            decrypted_result += chr(decrypted_char)
        except:
            print("Couln't add character: " + str(decrypted_char))
            exit(1)

    return noise_remover(decrypted_result)

def testing_failed():
    text = "Hello, World! This is a test text to check the encryption and decryption."
    key = "This is a key to test the encryption and decryption."
    if deuxrcrypt(uxrcrypt(text, key), key) != text:
        return True
    key = hash(key)
    if deuxrcrypt(uxrcrypt(text, key), key) != text:
        return True
    return False

if testing_failed():
    print("ENCRYPTION AND DECRYPTION DOESN'T WORK PROPERLY. FIX ultraxrcrypt.py!")
    exit(1)
else:
    print("ultraxrcrypt.py verified and ready to use.")

File: algorithm/ultraxrcrypt.js
const alphabet = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890";
const startMarker = "{~xr~}+";
const endMarker = "-{~xr~}";

// Custom modulo function to mimic Python's behavior
function mod(n, m) {
    return ((n % m) + m) % m;
}

// Hash function
function hash(key, customAlphabet, length = 64) {
    const alphabetToUse = customAlphabet || alphabet;
    key = key.toString();
    let result = "";
    let keysum = Array.from(key).reduce((acc, c) => acc + c.charCodeAt(0), 0);

    for (let i = 0; i < length; i++) {
        let keyitem = key.charCodeAt(i % key.length);

        let hasheditem = mod((keyitem * keyitem + Math.sin(keyitem * i * i - 29) * 110 + 401), 1000);
        hasheditem += mod((keyitem * 2 + Math.sin(keyitem * i * 91 - keysum * 7) * 160 + 301), 1000);
        hasheditem += mod((keyitem * 5 + Math.sin(i * 13 * keysum * 3) * 102 + 201), 777);
        hasheditem += Math.sin(keyitem * i + keysum * keyitem) * keysum + 101;
        hasheditem += keyitem * i + 77 * (i % key.length);

        let index = mod(Math.floor(hasheditem), alphabetToUse.length);
        result += alphabetToUse[index];
    }

    return result;
}

// Generate noise
function generateNoise(minLength = 1, maxLength = 16, context = alphabet) {
    if (context.length === 0) {
        context = alphabet;
    }

    let noise = "";
    let length = Math.floor(Math.random() * (maxLength - minLength + 1)) + minLength;
    for (let i = 0; i < length; i++) {
        noise += context[Math.floor(Math.random() * context.length)];
    }

    return startMarker + noise + endMarker;
}

// Add noise to text
function addNoise(text, count, minLength = 1, maxLength = 16, context = alphabet) {
    let selected = [];
    let textlen = text.length;

    for (let i = 0; i < count; i++) {
        selected.push(Math.floor(Math.random() * textlen));
    }
    selected.sort((a, b) => a - b);

    let result = "";
    let indexInSelected = 0;
    for (let i = 0; i < text.length; i++) {
        result += text[i];
        if (indexInSelected < selected.length && i === selected[indexInSelected]) {
            let noiseToAdd = generateNoise(minLength, maxLength, context);
            result += noiseToAdd;
            indexInSelected++;
        }
    }

    return result;
}

// Remove noise from text
function noiseRemover(text) {
    return text.replace(/\{~xr~\}\+.*?\-\{~xr~\}/g, '');
}

// Encryption function
function uxrcrypt(text, key, padding = 16) {
    let encryptedResult = "";

    // Add noise to the text
    text = addNoise(text, 5 + Math.floor(Math.random() * 11), 1, 16, alphabet);

    // Add padding noise
    text = generateNoise(1, padding, alphabet) + text + generateNoise(1, padding, alphabet);

    let keylen = key.length;
    let textlen = text.length;
    let keysum = Array.from(key).reduce((acc, c) => acc + c.charCodeAt(0), 0);

    for (let i = 0; i < textlen; i++) {
        let keyitem = key.charCodeAt(i % keylen);
        let encryptedChar = text.charCodeAt(i);

        // Encryption formula
        encryptedChar += mod((keylen * keylen + keyitem * keyitem), 111);
        encryptedChar += mod((37 * keyitem), 31);
        encryptedChar += mod(((textlen * textlen + 7) * keylen), 222);
        encryptedChar += mod((keylen * textlen - 1), 333);
        encryptedChar += mod((keysum * textlen + keysum * keylen - (keysum ^ keyitem)), 255);
        encryptedChar -= mod(((2 * textlen) % 8 + 1), 33);
        encryptedChar += mod(((i % keylen * 3) % 37), 555);
        encryptedChar += ((keyitem * 11) % 5) * 1001 + 777;

        encryptedResult += String.fromCharCode(encryptedChar);
    }

    return encryptedResult;
}

// Decryption function
function deuxrcrypt(text, key) {
    let decryptedResult = "";

    let keylen = key.length;
    let textlen = text.length;
    let keysum = Array.from(key).reduce((acc, c) => acc + c.charCodeAt(0), 0);

    for (let i = 0; i < textlen; i++) {
        let keyitem = key.charCodeAt(i % keylen);
        let decryptedChar = text.charCodeAt(i);

        // Decryption formula (reversing encryption logic)
        decryptedChar -= ((keyitem * 11) % 5) * 1001 + 777;
        decryptedChar -= mod(((i % keylen * 3) % 37), 555);
        decryptedChar += mod(((2 * textlen) % 8 + 1), 33);
        decryptedChar -= mod((keysum * textlen + keysum * keylen - (keysum ^ keyitem)), 255);
        decryptedChar -= mod((keylen * textlen - 1), 333);
        decryptedChar -= mod(((textlen * textlen + 7) * keylen), 222);
        decryptedChar -= mod((37 * keyitem), 31);
        decryptedChar -= mod((keylen * keylen + keyitem * keyitem), 111);

        decryptedResult += String.fromCharCode(decryptedChar);
    }

    // Remove noise from the decrypted text
    return noiseRemover(decryptedResult);
}

// Testing function
function testingFailed() {
    const text = "Hello, World! This is a test text to check the encryption and decryption.";
    let key = "This is a key to test the encryption and decryption.";

    // Test with original key
    let encryptedText = uxrcrypt(text, key);
    let decryptedText = deuxrcrypt(encryptedText, key);

    if (decryptedText !== text) {
        return true;
    }

    // Test with hashed key
    key = hash(key);
    encryptedText = uxrcrypt(text, key);
    decryptedText = deuxrcrypt(encryptedText, key);

    if (decryptedText !== text) {
        return true;
    }

    return false;
}

// Run the test
if (testingFailed()) {
    console.error("ENCRYPTION AND DECRYPTION DOESN'T WORK PROPERLY. FIX ultraxrcrypt.js!");
    throw new Error("Encryption/Decryption failed.");
} else {
    console.log("ultraxrcrypt.js verified and ready to use.");
}
